relayer/src/foreign_client.rs

```diff
@@ -406,6 +476,232 @@ impl ForeignClient {
 
         Ok(())
     }
+
+    pub fn update_client_event(
+        &self,
+        consensus_height: Height,
+    ) -> Result<Option<UpdateClient>, ForeignClientError> {
+        let request = QueryClientEventRequest {
+            height: Height::zero(),
+            event_id: IbcEventType::UpdateClient,
+            client_id: self.id.clone(),
+            consensus_height,
+        };
+
+        // Tx query fails if we don't wait a bit here (block is not ready/ indexed?)
+        thread::sleep(Duration::from_millis(100));
+        let res = self
+            .dst_chain
+            .query_txs(QueryTxRequest::Client(request))
+            .map_err(|e| {
+                ForeignClientError::Misbehaviour(format!(
+                    "failed to query Tx-es for update client event {}",
+                    e
+                ))
+            })?;
+
+        if res.is_empty() {
+            return Ok(None);
+        };
+
+        assert_eq!(res.len(), 1);
+        let event = res[0].clone();
+        let update = downcast!(event.clone() => IbcEvent::UpdateClient).ok_or_else(|| {
+            ForeignClientError::Misbehaviour(format!(
+                "query Tx-es returned unexpected event {}",
+                event.to_json()
+            ))
+        })?;
+        Ok(Some(update))
+    }
+
+    /// Retrieves all consensus states for this client and sorts them in reverse height
+    /// order. If consensus states are not pruned on chain, then last consensus state is the one
+    /// installed by the `CreateClient` operation.
+    fn consensus_states(&self) -> Result<Vec<AnyConsensusStateWithHeight>, ForeignClientError> {
+        let mut consensus_states = self
+            .dst_chain
+            .query_consensus_states(QueryConsensusStatesRequest {
+                client_id: self.id.to_string(),
+                pagination: None,
```

Comment:
Do we know what the semantics of setting `pagination` to `None` are? The Protobuf definition for `PageRequest` says (emphasis mine):

>     /// limit is the total number of results to be returned in the result page.
>     /// **If left empty it will default to a value to be set by each app.**

which leads me to think that we might not get all values at once. Perhaps we should always set it to something like `u64::MAX` to be sure we get them all.

https://github.com/informalsystems/ibc-rs/blob/a11c3c0847677280408a518dc353c89612c3054d/proto/src/prost/cosmos.base.query.v1beta1.rs#L20-L23

Reply:
Yes this is an issue, please see #811